plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/HiveFileIterator.java

```diff
@@ -75,63 +79,102 @@ public HiveFileIterator(
-                }
+        if (remoteIterator == null) {
+            // First iteration, initialize the iterator
+            remoteIterator = getLocatedFileStatusRemoteIterator(path);
+        }
 
-                if (status.isDirectory()) {
-                    switch (nestedDirectoryPolicy) {
-                        case IGNORED:
-                            continue;
-                        case RECURSE:
-                            paths.add(status.getPath());
-                            continue;
-                        case FAIL:
-                            throw new NestedDirectoryNotAllowedException(status.getPath());
-                    }
-                }
+        String pathPrefix = path.toUri().getPath();
+        while (remoteIterator.hasNext()) {
+            LocatedFileStatus status = getLocatedFileStatus(remoteIterator);
 
-                return status;
+            // Ignore hidden files and directories
+            if (nestedDirectoryPolicy == NestedDirectoryPolicy.RECURSE && isHiddenOrWithinHiddenParentDirectory(status.getPath(), pathPrefix)) {
+                continue;
+            }
+            else if (isHiddenFileOrDirectory(status.getPath())) {
+                continue;
             }
 
-            if (paths.isEmpty()) {
-                return endOfData();
+            if (status.isDirectory()) {
+                switch (nestedDirectoryPolicy) {
+                    case IGNORED:
+                        continue;
+                    case RECURSE:
+                        // Recursive listings call listFiles which should not return directories, this is a contract violation
+                        // and can be handled the same way as the FAIL case
+                    case FAIL:
+                        throw new NestedDirectoryNotAllowedException(status.getPath());
+                }
             }
-            remoteIterator = getLocatedFileStatusRemoteIterator(paths.removeFirst());
+            return status;
         }
+
+        return endOfData();
     }
 
     private Iterator<LocatedFileStatus> getLocatedFileStatusRemoteIterator(Path path)
     {
         try (TimeStat.BlockTimer ignored = namenodeStats.getListLocatedStatus().time()) {
-            if (ignoreAbsentPartitions && !exists(path)) {
-                return emptyIterator();
+            return new FileStatusIterator(table, path, fileSystem, directoryLister, namenodeStats, nestedDirectoryPolicy == NestedDirectoryPolicy.RECURSE);
+        }
+        catch (TrinoException e) {
+            if (ignoreAbsentPartitions) {
+                try {
+                    if (!fileSystem.exists(path)) {
+                        return emptyIterator();
+                    }
+                }
+                catch (Exception ee) {
+                    TrinoException trinoException = new TrinoException(HIVE_FILESYSTEM_ERROR, "Failed to check if path exists: " + path, ee);
+                    trinoException.addSuppressed(e);
```

Comment:
It thought it's usually the other way around: e.g.: `e.addSuppressed(ee)` as the original exception `e` is more important? (it is also required to check if `e != ee` as self suppression is not allowed)

Reply:
Usually, that's true- but in this case since I was preserving the observable exception order that was present before, ie: if the preemptive check failed before then this is the exception that you would receive.

After this change, the check only occurs after the listing fails, so the question is which exception should take precedence. Since we're in the path of "ignore absent partitions" it seems like we want to _mostly_ ignore the original failure (although keep it as a suppressed exception).

On the other point, we know that `trinoException != ee` because we instantiated it directly ourselves.